File: blazorWasm/Workers/QRGenerator.razor.js
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
import { dotnet, exit } from '../_framework/dotnet.js'

let assemblyExports = null;

self.addEventListener('message', async function(e) {
    switch (e.data.command)
    {
        case "startDotnet":
            console.log("Starting dotnet, says the worker");
            await startDotnet();
            break;
        case "generateQRCode":
            if (!assemblyExports)
                throw new Error("Exports not found");
            const size = Number(e.data.size);
            const text = e.data.text;
            if (size === undefined || text === undefined)
                new Error("Inner error, got empty QR generation data from React");
            const imageBytes = assemblyExports.QRGenerator.Generate(text, size);
            self.postMessage({ command: "generateQRCodeResponse", image: imageBytes.buffer }, [imageBytes.buffer]);
        default:
            self.postMessage(e.data.command);
            break;
    }
    }, false);

async function startDotnet(){
    try {
        self.postMessage("creating dotnet");
        const { setModuleImports, getAssemblyExports, getConfig } = await dotnet
            .create();
        self.postMessage("getting config");
        const config = getConfig();
        self.postMessage("getting exports");
        assemblyExports = await getAssemblyExports(config.mainAssemblyName);
    
        self.postMessage("setting imports");
        setModuleImports("worker.razor.js", {
            QRGenerator: {
                sendErrorMessage
            }
        });
    
        self.postMessage({ command: "exportsReady" });
        self.postMessage("starting dotnet");
        await dotnet.run();
    }
    catch (err) {
        console.log(`err: ${err}; ${err.message}`);
        exit(2, err);
    }
}

export function sendErrorMessage(message) {
    console.log(`sendErrorMessage is passing message: ${message}`);
    self.postMessage({ command: "error", message: message });
}

File: blazorWasm/Workers/QRGenerator.razor.cs
// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
using System;
using System.Runtime.InteropServices.JavaScript;
using System.Runtime.Versioning;
using QRCoder;

// https://github.com/codebude/QRCoder

[SupportedOSPlatform("browser")]
public partial class QRGenerator
{
    private static readonly int MAX_QR_SIZE = 20;

    [JSExport]
    internal static byte[] Generate(string text, int qrSize)
    {
        if (qrSize >= MAX_QR_SIZE)
        {
            SendErrorMessage($"QR code size must be less than {MAX_QR_SIZE}. Try again.");
            return Array.Empty<byte>();
        }
        QRCodeGenerator qrGenerator = new QRCodeGenerator();
        QRCodeData qrCodeData = qrGenerator.CreateQrCode(text, QRCodeGenerator.ECCLevel.Q);
        BitmapByteQRCode qrCode = new BitmapByteQRCode(qrCodeData);
        return qrCode.GetGraphic(qrSize);
    }

    [JSImport("QRGenerator.sendErrorMessage", "worker.razor.js")]
    internal static partial void SendErrorMessage(string message);
}
